src/optics.jl

```diff
@@ -318,6 +337,189 @@ function _modify(f, obj, r::Recursive, ::ModifyBased)
+
+julia> obj = (a=missing, b=1, c=(d=missing, e=(f=missing, g=2)))
+(a = missing, b = 1, c = (d = missing, e = (f = missing, g = 2)))
+
+julia> set(obj, Query(ismissing), (1.0, 2.0, 3.0))
+(a = 1.0, b = 1, c = (d = 2.0, e = (f = 3.jjjjjjtk,rg, g = 2)))
+
+julia> obj = (1,2,(3,(4,5),6))
+(1, 2, (3, (4, 5), 6))
+
+julia> modify(x -> 100x, obj, Recursive(x -> (x isa Tuple), Elements()))
+(100, 200, (300, (400, 500), 600))
+```
+$EXPERIMENTAL
+"""
+struct Query{Select,Descend,Optic<:Union{ComposedOptic,Properties}} <: AbstractQuery
+    select_condition::Select
+    descent_condition::Descend
+    optic::Optic
+end
+Query(select, descend = x -> true) = Query(select, descend, Properties())
+Query(; select=Any, descend=x -> true, optic=Properties()) = Query(select, descend, optic)
+
+OpticStyle(::Type{<:AbstractQuery}) = SetBased()
+
+struct ContextState{V,O,FN}
+    vals::V
+    obj::O
+    fn::FN
+end
+struct GetAllState{V}
+    vals::V
+end
+struct SetAllState{C,V,I}
+    change::C
+    vals::V
+    itr::I
+end
+
+const GetStates = Union{GetAllState,ContextState}
+
+@inline pop(x) = first(x), Base.tail(x)
+@inline push(x, val) = (x..., val)
+@inline push(x::GetAllState, val) = GetAllState(push(x.vals, val))
+@inline push(x::ContextState, val) = ContextState(push(x.vals, val), nothing, nothing)
+
+(q::Query)(obj) = getall(obj, q)
+
+getall(obj, q) = _getall(obj, q).vals
+function _getall(obj, q::Q) where Q<:Query
+    initial_state = GetAllState(())
+    _, final_state = let q=q
+        modify_stateful((obj, initial_state), q) do o, s
+            new_state = push(s, outer(q.optic, o, s))
+            o, new_state
+        end
+    end
+    final_state
+end
+
+function setall(obj, q::Q, vals) where Q<:Query
+    initial_state = SetAllState(Unchanged(), vals, 1)
+    final_obj, _ = let obj=obj, q=q, initial_state=initial_state
+        modify_stateful((obj, initial_state), q) do o, s
+            new_output = outer(q.optic, o, s)
+            new_state = SetAllState(Changed(), s.vals, s.itr + 1)
+            new_output, new_state
+        end
+    end
+    return final_obj
+end
+
+function context(f::F, obj, q::Q) where {F,Q<:Query}
+    initial_state = ContextState((), nothing, nothing)
+    _, final_state = let f=f
+        modify_stateful((obj, initial_state), q) do o, s
+            new_state = push(s, f(s.obj, known(s.fn)))
+            o, new_state
+        end
+    end
+    return final_state.vals
+end
+
+modify(f, obj, q::Query) = setall(obj, q, map(f, getall(obj, q)))
+
+@inline function modify_stateful(f::F, (obj, state), q::Q) where {F,Q<:Query}
+    let f=f, q=q
+        modify_stateful((obj, state), inner(q.optic)) do o, s
+            if (q::Q).select_condition(o)
+                (f::F)(o, s)
+            elseif (q::Q).descent_condition(o)
+                ds = descent_state(s)
+                o, ns = modify_stateful(f::F, (o, ds), q::Q)
+                o, merge_state(ds, ns)
+            else
+                o, s
+            end
```

Comment:
Thanks a lot for pushing this further. I also played around with implementing this, see this branch:
https://github.com/JuliaObjects/Accessors.jl/tree/modify_stateful  
Like with your implementation I encountered inference issues. I do not fully understand them, but it seems the compiler dislikes this switch. To explore what's going on I replaced `elseif (q::Q).descent_condition(o)` into  `elseif true` but the thing would still not infer. OTOH manually deleting the `elseif` branch did help the compiler to infer.

Note I did this experimentation with my branch, which is somewhat different then yours, but I guess has the same underlying issues. Worth checking if things are better with the new ConstructionBase version and nightly julia.

If not I think it would be good to produce an MWE and report it upstream. I don't know if it is considered a bug, but maybe we'll get an explanation of why it is this way and how to work around it.

A quick and dirty workaround would be to introduce a more narrow query where you can only select and descent based on type. Then the `if elseif ...` could be replaced by three methods.

Reply:
Ok interesting we both hit the same problem. 

I think the differences are mostly just my code to handle context and optional construction in setall. But those aren't whats causing the instability... 

It seems like the problem is calling `modify_stateful` recursively? It still seems broken even without the `elseif`, and the `elseif` was fine in the previous version. I don't understand exactly what is different between these implementations and my previous one or Flatten.jl, but it could be the additional complexity of the return objects? Or there is a little more happening at runtime in `modify_stateful` than I had in `mapobject` or Flatten.jl, and the compiler is giving up on something?

I also saw no difference between ConstructionBase versions.

Reply:
Seems its the layers of anonymous functions. Flatten.jl has the [shared code](https://github.com/rafaqz/Flatten.jl/blob/master/src/Flatten.jl#L29-L33) inside the generated functions, and simple [separate generated functions](https://github.com/rafaqz/Flatten.jl/blob/master/src/Flatten.jl#L124-L131) for getall/setall (flatten/reconstruct).

I much prefer the clarity of the current solution here. But it seems like if we want this to compile away we cant use anonymous functions inside the recursion.

Reply:
It does not seem to be related to anonymous functions per see. I played with eliminating all anonymous functions.
I replaced code like
```julia
modify_stateful(...) do os
    do_stuff(os, captured, variables)
end
```
with code like
```julia
struct DoStuff{C,V}
    captured::C
    variables::V
end
(f::DoStruff)(os) = do_stuff(os, f.captured, f.variables)
modify_stateful(DoStuff(captured, variables), ...)
```
But it helped in very small examples, small examples still fail. Also, `@inline` everything did not help.

Reply:
Damn, that's exhaustive.

It feels like the compiler is using some heuristic to give up on solving the returned state of the recursion instead of just trying. A single non-recursive `modify_stateful` compiles away completely, but it wont if called recursively.

Reply:
Ok, think i found the issue. If you define the select and descent conditions like this it's type stable:

```julia
select_condition(x::NamedTuple) = true
select_condition(x) = false

descent_condition(x::Tuple) = true
descent_condition(x) = false
```

So the compiler optimises these methods differently to methods that are passed in in any way. Even with the function in the  type instead of fields, or just passing a Type to a 2 arg version of `descend_condition` still doesn't work. It's just giving up and returning `Any`. If you actually include the types to descend on in the code, it's type-stable. 

I had assumed previously that using a type from a type parameters was practically the same as writing the type in the code manually, but it seems like it isn't.

Reply:
Awesome that you found the issue! This looks like a bug to me, can you maybe produce a MWE and report it upstream?

Reply:
Yeah I'm cutting it down to size, but its probably 1.6 inference problems like JuliaLang/julia#35800.

It's all type stable on 1.4 - that's why this feels weird for performance intuition. These things used to just work.

Reply:
Eh only some things are more stable on 1.4. #35800 just has me not trusting the compiler anymore.